diff --git a/momo/Makefile b/momo/Makefile
index 23aef7c..34556fc 100644
--- a/momo/Makefile
+++ b/momo/Makefile
@@ -4,17 +4,36 @@ PKG_NAME:=momo
 PKG_VERSION:=2026.04.09
 PKG_RELEASE:=1
 
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.gz
+PKG_SOURCE_SUBDIR:=$(PKG_NAME)-$(PKG_VERSION)
+PKG_SOURCE_PROTO:=git
+PKG_SOURCE_URL:=https://github.com/reF1nd/sing-box.git
+PKG_SOURCE_VERSION:=reF1nd-testing
+PKG_MIRROR_HASH:=skip
+
+PKG_LATEST_TAG:=$(shell curl -sL https://api.github.com/repos/reF1nd/sing-box/tags | jq -r '.[].name | select(contains("reF1nd"))' | head -n 1 | sed 's/^v//')
+
 PKG_LICENSE:=GPL-3.0+
 PKG_MAINTAINER:=Joseph Mory <morytyann@gmail.com>
 
+PKG_BUILD_DEPENDS:=golang/host
+PKG_BUILD_PARALLEL:=1
+PKG_BUILD_FLAGS:=no-mips16
+
+GO_PKG:=github.com/sagernet/sing-box
+GO_PKG_BUILD_PKG:=$(GO_PKG)/cmd/sing-box
+GO_PKG_LDFLAGS_X:=$(GO_PKG)/constant.Version=$(PKG_LATEST_TAG)
+GO_PKG_TAGS:=with_quic,with_utls,with_gvisor,with_clash_api,with_musl
+
 include $(INCLUDE_DIR)/package.mk
+include $(TOPDIR)/feeds/packages/lang/golang/golang-package.mk
 
 define Package/momo
   SECTION:=net
   CATEGORY:=Network
   TITLE:=Transparent Proxy with sing-box on OpenWrt.
   URL:=https://github.com/nikkinikki-org
-  DEPENDS:=+ca-bundle +curl firewall4 +ip-full +kmod-inet-diag +kmod-nft-socket +kmod-nft-tproxy +kmod-tun +kmod-dummy +sing-box
+  DEPENDS:=+ca-bundle +curl firewall4 +ip-full +kmod-inet-diag +kmod-nft-socket +kmod-nft-tproxy +kmod-tun +kmod-dummy
 endef
 
 define Package/momo/conffiles
@@ -22,6 +41,7 @@ define Package/momo/conffiles
 endef
 
 define Package/momo/install
+	$(call GoPackage/Package/Install/Bin,$(1))
 	$(INSTALL_DIR) $(1)/etc/momo
 	$(INSTALL_DIR) $(1)/etc/momo/ucode
 	$(INSTALL_DIR) $(1)/etc/momo/scripts
@@ -70,4 +90,5 @@ define Build/Compile
 
 endef
 
+$(eval $(call GoBinPackage,momo))
 $(eval $(call BuildPackage,momo))
\ No newline at end of file
